src/ui.rs

```diff
@@ -3047,40 +3047,60 @@ impl Component for SectionLineView<'_> {
+
+            let mut last_index = 0;
+            // Find index of the start of each character to replace
+            for (idx, char) in line.match_indices(|char| replacement_characters(char).is_some()) {
+                // Push the string leading up to the character and the styled replacment string
+                if let Some(replacment_string) =
+                    char.chars().next().and_then(replacement_characters)
+                {
+                    spans.push(Span::raw(&line[last_index..idx]));
+                    spans.push(Span::styled(replacment_string, CONTROL_CHARACTER_STYLE));
+                    // Move the "cursor" to just after the character we're replacing
+                    last_index = idx + char.len();
+                }
+            }
+            // Append anything remaining after the last replacment
+            let remaining_line = &line[last_index..];
+            if !remaining_line.is_empty() {
+                spans.push(Span::raw(remaining_line));
+            }
+        }
+
         viewport.draw_blank(Rect {
             x: viewport.mask_rect().x,
             y,
             width: viewport.mask_rect().width,
             height: 1,
         });
-        match inner {
+
+        match &self.inner {
             SectionLineViewInner::Unchanged { line, line_num } => {
-                let style = Style::default().add_modifier(Modifier::DIM);
                 // Pad the number in 5 columns because that will align the
                 // beginning of the actual text with the `+`/`-` of the changed
                 // lines.
-                let line_num_rect =
-                    viewport.draw_span(x, y, &Span::styled(format!("{line_num:5} "), style));
-                let (line, line_end) = match line.strip_suffix('\n') {
-                    Some(line) => (
-                        Span::styled(line, style),
-                        Some(Span::styled(
-                            NEWLINE_ICON,
-                            Style::default().fg(Color::DarkGray),
-                        )),
-                    ),
-                    None => (Span::styled(*line, style), None),
-                };
-                let line_rect = viewport.draw_text(
-                    line_num_rect.x + line_num_rect.width.unwrap_isize(),
-                    line_num_rect.y,
-                    &line,
-                );
-                if let Some(line_end) = line_end {
-                    viewport.draw_span(line_rect.x + line_rect.width.unwrap_isize(), y, &line_end);
-                }
+                let line_number = Span::raw(format!("{line_num:5} "));
+                let mut spans = vec![line_number];
+                push_spans_from_line(line, &mut spans);
+
+                const UI_UNCHANGED_STYLE: Style = Style::new().add_modifier(Modifier::DIM);
+                viewport.draw_line(x, y, &Line::from(spans).style(UI_UNCHANGED_STYLE));
```

Comment:
note: I considered using `Span::styled` for `line_number` like it used to be, but I'm interested in if `Line::from(spans).style(UI_UNCHANGED_STYLE)` is enough and that we don't need to specify the style twice.

Reply:
As far as I can tell it should be enough and we shouldn't need to specify the style twice.